### ISSUE TITLE
Intelligent Amend Commit

### ISSUE THREAD

**Issue by NJAldwin** [gitextensions/gitextensions#565]
Currently, it's rather easy to mess up when doing an "amend commit".  First, one must click the **Amend Commit** button in the commit dialog, in order to load the last commit & commit message.  Then, when one is done, one must click the **Amend Commit** button _again_---**_NOT**_ the **Commit** button---to finish amending the commit.

Perhaps, to make this more intuitive, we could have a message or notification that comes up when someone clicks the **Commit** button after the **Amend Commit** button which asks them if they meant to amend the commit by pressing the **Amend Commit** button again.

This probably isn't the best way to solve the problem.  However, I do think that it needs a better solution.  Right now it is counter-intuitive.

**Comment by patricktokeeffe** [comment]
Things have improved since this was opened. The current amend checkbox + single commit button works great for me. Thanks everyone who made that happen. 

And I agree 100% with @jasir: the gitgui behavior of loading `HEAD^` back into the staged area when amending is pure gold. I look forward to that happening with GitEx too.

**Comment by sadastronaut** [comment]
"Amend Commit" and the Commit dialog, in general, for me has several issues.  

1) I can only populate the commit dialog with the last commit if I clear out all the text in the commit message, then clikc Amend Commit. 
2)  When I do the Commit after Step 1, it pops up the warning saying I'm going to rewrite history and all that.  If I do that, it clears out my commit dialog and asks me to please enter a commit message.  Perhaps the fact that I'm using a template messes it up.  

Unrelated to the commits...if I have modified files but I forget to stage them, I enter in my commit message, hit the commit button, confirm that I want to stage and commit all changes via the commit dialog, then it clears out the commit message and asks me to enter a commit message.  Again, this could be related to my template (not done via extensions but a COMMITMESSAGE.txt handled by git itself).

**Comment by ghanique** [comment]
I don't have any issues with the Commit Message textbox being cleared, ever.

**Comment by mrexodia** [comment]
I have been looking into this a little bit, the checkbox for amending a commit works great, but the downside is that you cannot really undo changes that were made in the commit. The Git GUI (which is rather terrible) handles this nicely by including the changes from the last commit in the staging area:

![](https://i.imgur.com/9v7SpRU.png)

Probably introducing this is a non-trivial change but I think it could be very useful!

**Comment by johndog** [comment]
Two issues, both of which have been mentioned:
- If there is text in the commit message window already, clicking amend won't clear it and replace it with the prior message.  This is useful sometimes, but more often it's a worry that I have the wrong message, so I end up clearing the text, then clearing/checking the amend box again to bring it back.  
  - **Suggestion A**:  Color the message area somehow (text or background) to show that the text currently disagrees with the message at HEAD
  - **Suggestion B**:  When clicking amend checkbox, if the text is different than the commit message for HEAD, always save it away, clear it and replace it with the text from HEAD.  Then when clicking "commit", restore the saved text.  This would accelerate a bunch of my workflows where I suspend a commit in progress (and its description) in order to amend the previous commit, and then resume.
- It's hard to review the amended commit in whole before commiting; it's hard to undo changes made in the HEAD commit.  
  - **Suggestion**: Provide a toggle that changes the index view from being relative to HEAD^ to relative to HEAD.  There are pros and cons to both views.

**Comment by GintasS** [comment]
@RussKie this is a duplicate, a new issue exists #2260 with recent activity.

**Comment by johndog** [comment]
#565 concerns the change description (which already does load, but only if the field is currently empty), #2260 concerns the modified file staging UI.  If #2260 is reframed as "holistic experience when checking the amend box", it can subsume both.

**Comment by gerhardol** [comment]
Duplicate of #2260
The "amend experience" to be handled in one issue.
